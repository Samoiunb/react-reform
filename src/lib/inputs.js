import wrapInput from "../wrap-input";

export const Text = wrapInput("Text", "input", {defaultProps: {type: "text"}});

export const Checkbox = wrapInput("Checkbox", "input", {
  defaultProps: {type: "checkbox"},
  uncontrolled: {
    extractValue: node => node.checked,
    setValue: (node, value) => node.checked = !!value
  }
});
