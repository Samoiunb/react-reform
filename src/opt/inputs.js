import wrapInput from "../core/wrap-input";

export const Text = wrapInput("Text", "input", {
  defaultProps: {type: "text"},
  extractValueFromOnChange: e => e.target.value
});

export const Textarea = wrapInput("Textarea", "textarea", {
  extractValueFromOnChange: e => e.target.value
});

export const Password = wrapInput("Password", "input", {
  defaultProps: {type: "password"},
  extractValueFromOnChange: e => e.target.value
});

export const Select = wrapInput("Select", "select", {
  extractValueFromOnChange: e => e.target.value
});

export const Checkbox = wrapInput("Checkbox", "input", {
  defaultProps: {type: "checkbox"},
  extractValueFromOnChange: e => e.target.checked,
  valueToProps: value => ({checked: value})
});
