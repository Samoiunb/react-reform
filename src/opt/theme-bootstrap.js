import React from "react";
import classNames from "classnames";

export default (FormContainer, Fields, {globalErrors, submitForm}) => (
  <FormContainer>
    <Fields>
      {/* validations = [{type: "required", isValid: true|false|"pending", hintMessage: "is required", errorMessage: "..."}] */}
      {(Field, {label, type, validations, isTouched, isFocused, hasFailedToSubmit, fieldProps, id}) => {
        const hasError = validations.some(({isValid}) => isValid === false);
        const showError = (isTouched || hasFailedToSubmit) && !isFocused && hasError;
        const showWarning = isFocused && hasError;

        const validationList = validations
          .filter(({isValid}) => fieldProps.showHints || ((isTouched || hasFailedToSubmit) && isValid !== true))
          .map(({isValid, errorMessage, hintMessage, type: valType}) => {
            if (isValid === false) return {message: errorMessage, valType};
            if (isValid === true) return {message: hintMessage, valType};
            return {message: isValid, valType};
          })
          .map(({valType, message}) => <span key={valType} className="help-block">{message}</span>);

        if (type === "Checkbox") {
          return (
            <div className={classNames("checkbox", {"has-error": showError, "has-warning": showWarning})}>
              <label htmlFor={id}>
                <Field id={id}/> {label}
              </label>
              {validationList}
            </div>
          );
        } else {
          return (
            <div className={classNames("form-group", {"has-error": showError, "has-warning": showWarning})}>
              <label className="control-label" htmlFor={id}>{label}</label>
              <Field className="form-control" id={id}/>
              {validationList}
            </div>
          );
        }
      }}
    </Fields>
    <footer>
      {globalErrors.length ? (
        globalErrors.map((error, i) => (
          <div key={i} className="alert alert-danger">{error}</div>
        ))
      ) : null}
      <button className="btn btn-default">Submit</button>
    </footer>
  </FormContainer>
);
