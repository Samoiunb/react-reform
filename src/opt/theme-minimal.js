import React from "react";

export default (FormContainer, Fields, {globalErrors, submitForm, status}) => (
  <FormContainer>
    {globalErrors.length ? (
      globalErrors.map((error, i) => (
        <div key={i} style={{color: "red"}}>{error}</div>
      ))
    ) : null}
    <Fields>
      {(Field, {label, validations, isTouched, fieldProps, id}) => {
        const validationList = validations
          .filter(({isValid}) => fieldProps.showHints || ((isTouched || status === "preSubmitFail") && isValid !== true))
          .map(({isValid, errorMessage, hintMessage, type}) => {
            if (isValid === false) return {message: errorMessage, type};
            if (isValid === true) return {message: hintMessage, type};
            return {message: isValid, type};
          })
          .map(({type, message}) => <span key={type} style={{color: "red"}}>{message}</span>);
        return (
          <div>
            <label htmlFor={id}>{label}</label>
            <Field id={id}/>
            {validationList}
          </div>
        );
      }}
    </Fields>
    <button>Submit</button>
  </FormContainer>
);
