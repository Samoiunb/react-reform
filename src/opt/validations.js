// <Text is-required/>

// taken from here: http://stackoverflow.com/a/1373724/616974
const emailRegex = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/

export default {
  required: {
    isValid: val => val === 0 || (!!val && (typeof val !== 'string' || val.trim().length > 0)),
    errorMessage: () => 'is required',
    hintMessage: () => 'is required'
  },

  // <Text is-email/>
  email: {
    isValid: val => emailRegex.test(val),
    errorMessage: () => 'is not a valid email address',
    hintMessage: () => 'needs to be an email'
  },

  // <Text has-minlength={5}/>
  minlength: {
    isValid: (val, {arg}) => (val || '').toString().length >= arg,
    errorMessage: (val, {arg}) => {
      const currLength = (val || '').toString().length
      return `minimal length: ${currLength}/${arg}`
    }
  },

  // <Text has-maxlength={5}/>
  maxlength: {
    isValid: (val, {arg}) => (val || '').toString().length <= arg,
    errorMessage: (val, {arg}) => {
      const currLength = (val || '').toString().length
      return `maximal length: ${currLength}/${arg}`
    }
  },


  // <Text has-pattern={/^\d+(\.\d+)?$/}/>
  pattern: {
    isValid: (val, {arg}) => (typeof arg === 'string' ? new RegExp(arg) : arg).test(val),
    errorMessage: val => `'${val}' is not valid`,
    hintMessage: (val, {arg}) => `needs to correspond to this pattern: ${arg}`
  }
}
