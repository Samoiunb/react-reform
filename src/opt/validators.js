import {registerValidator} from "../validator-store";

// <Text is-required/>
registerValidator("required", {
  isValid: val => val === 0 || !!val && (typeof val !== "string" || val.trim().length > 0),
  errorMessage: () => "is required",
  hintMessage: () => "is required"
});

// taken from here: http://stackoverflow.com/a/1373724/616974
const emailRegex = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;

// <Text is-email/>
registerValidator("email", {
  isValid: val => emailRegex.test(val),
  errorMessage: () => `is not a valid email address`,
  hintMessage: () => "needs to be an email"
});

// <Text has-minlength={5}/>
registerValidator("minlength", {
  isValid: (val, opts) => (val || "").toString().length >= opts,
  errorMessage: (val, opts) => {
    const currLength = (val || "").toString().length;
    return `minimal length: ${currLength}/${opts}`;
  }
});

// <Text has-maxlength={5}/>
registerValidator("maxlength", {
  isValid: (val, opts) => (val || "").toString().length <= opts,
  errorMessage: (val, opts) => {
    const currLength = (val || "").toString().length;
    return `maximal length: ${currLength}/${opts}`;
  }
});


// <Text has-pattern={/^\d+(\.\d+)?$/}/>
registerValidator("pattern", {
  isValid: (val, opts) => (typeof opts === "string" ? new RegExp(opts) : opts).test(val),
  errorMessage: val => `'${val}' is not valid`,
  hintMessage: (val, opts) => `needs to correspond to this pattern: ${opts}`
});

registerValidator("dummy-unique", () => { // to be moved to example section
  const data = {};
  return {
    isValid: (val, ctx, askAgain) => {
      if (data[val] === undefined) {
        data[val] = true;
        setTimeout(askAgain, 1000);
        return "pending";
      }
      return data[val];
    },
    errorMessage: val => `'${val}' is not unique`,
    hintMessage: () => "needs to be unique"
  };
});
