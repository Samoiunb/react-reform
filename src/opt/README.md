These modules are optional and can be required separatedly.

to use them simply import them like this:

```
// ES6 Style
import {Text, Checkbox, Password, Textarea, Select} from "react-themed-forms/opt/inputs";
```

or


```
// classic
var Text = require("react-themed-forms/opt/inputs").Text;
```
