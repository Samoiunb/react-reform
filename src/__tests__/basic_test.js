import renderer from 'react-test-renderer'
import React from 'react'
import {Form, ReformContext, createTheme} from '../index'
import defaultValidations from '../opt/validations'
import {Text} from '../opt/inputs'

test('It doesn\'t explode when rendering a basic form', () => {

  const defaultTheme = createTheme({
    renderForm: (FormContainer, children, {directProps: {buttonLabel = 'Submit', ...remainingDirectProps}}) => (
      <FormContainer {...remainingDirectProps}>
        <div>{children}</div>
        <button>{buttonLabel}</button>
      </FormContainer>
    ),
    renderField: (Field, {directProps, name, validations, id}) => {
      const errors = validations
        .filter(({isValid}) => isValid === false)
        .map(({errorMessage, name}) => <span key={name}>{errorMessage} </span>)
      return (
        <div>
          <label htmlFor={id}>{name}</label>
          <Field id={id} {...directProps}/>
          {errors.length > 0 && <span>{errors}</span>}
        </div>
      )
    }
  })


  const component = renderer.create(
    <ReformContext themes={{default: defaultTheme}} validations={defaultValidations}>
      <div>
        <Form onSubmit={() => {}} buttonLabel="Custom button text">
          <Text name="field1" isRequired/>
        </Form>
      </div>
    </ReformContext>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
