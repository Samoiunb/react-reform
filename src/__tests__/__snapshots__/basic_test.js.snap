exports[`test It doesn't explode when rendering a basic form 1`] = `
<div>
  <form
    onSubmit={[Function]}>
    <div>
      <div>
        <label
          htmlFor="form1-field1">
          field1
        </label>
        <input
          id="form1-field1"
          name="field1"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          type="text"
          value="" />
        <span>
          <span>
            is required
             
          </span>
        </span>
      </div>
    </div>
    <button>
      Custom button text
    </button>
  </form>
</div>
`;
